File: scripts/perl/wrapper.pl
use strict;
use warnings;

use File::Basename;

use Cwd;
my $dotfiles_dir = Cwd::realpath(dirname(__FILE__) . '/../..');

require $dotfiles_dir . '/scripts/perl/shellquote.pl';

sub remove_path_from_env {
  my ($target) = @_;
  my @paths = split(/:/, $ENV{'PATH'});
  my @tmp = ();
  for my $path (@paths) {
    if ($path ne $target and $path ne "") {
      push @tmp, $path;
    }
  }
  $ENV{'PATH'} = join(':', @tmp);
}

sub print_and_exit {
  my ($cmdline, $code) = @_;

  if (open(SIGNAL, ">&=3")) {
    print SIGNAL String::ShellQuote::shell_quote(@$cmdline);
    close(SIGNAL);
    exit $code;
  } else {
    exec @$cmdline;
  }
}

sub run_original_if_required {
  my ($check_dash) = @_;
  if ($check_dash && scalar(@ARGV) && $ARGV[0] eq '-') {
    $ARGV[0] = basename($0);
    run_original(0);
  } elsif (!$check_dash && scalar(@ARGV)) {
    run_original();
  }
}

sub run_original {
  my ($add_program) = @_;

  if (!defined($add_program) || $add_program) {
    unshift @ARGV, basename($0);
  }
  print_and_exit(\@ARGV, 100);
}

1;

File: scripts/python/wrapper.py
#!/usr/bin/env python3
import os
import shlex
import sys


def remove_path_from_env(target):
    paths = os.environ.get('PATH', '').split(':')
    os.environ['PATH'] = ':'.join(p for p in paths if p and p != target)


def print_and_exit(cmdline, code):
    quoted = ' '.join(shlex.quote(str(a)) for a in cmdline)
    try:
        fd3 = os.fdopen(3, 'w')
        fd3.write(quoted)
        fd3.close()
        sys.exit(code)
    except OSError:
        os.execvp(cmdline[0], [str(a) for a in cmdline])


def run_original(add_program=True):
    args = list(sys.argv[1:])
    if add_program:
        args.insert(0, os.path.basename(sys.argv[0]))
    print_and_exit(args, 100)


def run_original_if_required(check_dash=False):
    if check_dash:
        if len(sys.argv) > 1 and sys.argv[1] == '-':
            sys.argv[1] = os.path.basename(sys.argv[0])
            run_original(add_program=False)
    else:
        if len(sys.argv) > 1:
            run_original()


if __name__ == '__main__':
    if len(sys.argv) < 2:
        print(f"Usage: {sys.argv[0]} <command> [args...]", file=sys.stderr)
        sys.exit(1)
    print_and_exit(sys.argv[1:], 100)
